### PR TITLE
Draft initial Python discovery sample gen

### PR DESCRIPTION
**Pull request by tcoffee-google**
Notes:
* modifies handling of Python imports to allow snippets to generate
  imports internally (similar to Java)
* does not handle media upload instances (to be resolved separately,
  along with snippets for other languages)
* relocates handling of exceptional inconsistencies to reflect patterns
  observed across Java, Python, and Node.js
* makes other minor corrections to existing sample gen code
* renames <lang>_discovery_gapic.yaml to <lang>_discovery.yaml for
  consistency with existing usage of 'gapic' vs. 'discovery' (correcting
  historical accident from renaming of 'veneer')

### CODE REVIEW

**Comment by tcoffee-google** [comment]
@vchudnov-g @pongad @saicheems

File: vgen/src/test/java/io/gapi/vgen/testdata/discoveries/nodejs/nodejs_sqladmin.v1beta4.json.baseline [hunk 1]
```diff
@@ -1331,20 +1331,13 @@ authFactory.getApplicationDefault(function(err, authClient) {
     auth: authClient
   };
 
-
-  var recur = function(err, result) {
+  sqladmin.users.list(request, function(err, result) {
```
Comment (pongad) [inline vgen/src/test/java/io/gapi/vgen/testdata/discoveries/nodejs/nodejs_sqladmin.v1beta4.json.baseline]: This new result is correct. The NodeJS API seems to think this is not page streaming. It is rather unfortunate that the compilecheck didn't catch this. I'll open an issue.

File: vgen/src/main/java/io/gapi/vgen/py/PythonDiscoveryContext.java [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+          .put("float", "0.0")
+          .build();
+
+  /**
+   * A map from {@link ApiaryConfig#stringFormat} label (or null) to corresponding default value.
+   */
+  private static final ImmutableMap<String, String> STRING_DEFAULT_MAP =
+      ImmutableMap.<String, String>builder()
+          .put("byte", "''"
+              + "  # base64-encoded string of bytes: see http://tools.ietf.org/html/rfc4648 ")
```
Comment (pongad) [inline vgen/src/main/java/io/gapi/vgen/py/PythonDiscoveryContext.java]: Remove trailing space in string.

File: vgen/src/main/java/io/gapi/vgen/py/PythonDiscoveryContext.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+              + "  # 'YYYY-MM-DD': see datetime.date.isoformat()")
+          .put("date-time", "'" + new DateTime(0L).toStringRfc3339() + "'"
+              + "  # 'YYYY-MM-DDThh:mm:ss.fffZ' (UTC): see datetime.datetime.isoformat()")
+          .build();
+
+  private static final ImmutableMap<String, String> RENAMED_METHOD_MAP =
+      ImmutableMap.<String, String>builder()
+          .build();
+
+  private PythonContextCommon pythonCommon;
```
Comment (pongad) [inline vgen/src/main/java/io/gapi/vgen/py/PythonDiscoveryContext.java]: Can this be made final?

File: vgen/src/test/java/io/gapi/vgen/testdata/discoveries/py/py_appengine.v1beta4.json.baseline [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+service = discovery.build('appengine', 'v1beta4', credentials=credentials)
+
+
+# TODO: Change placeholders below to values for parameters to the 'list' method:
+# * Part of `name`. Name of the resource requested. For example: "apps/myapp".
+appsId = ''
+
+modules = service.apps().modules()
+request = modules.list(appsId)
+while request is not None:
```
Comment (pongad) [inline vgen/src/test/java/io/gapi/vgen/testdata/discoveries/py/py_appengine.v1beta4.json.baseline]: Python might prefer `while request:`

File: vgen/src/main/resources/io/gapi/vgen/py/discovery_fragment.snip [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    @end
+@end
+
+@private TODO()
+    {@context.getTODO()}
+@end
+
+@private alwaysImport() fill
+  {@context.silent(importHandler.addImportLocal("apiclient", "discovery"))}
+  {@context.silent(importHandler.addImportLocal("oauth2client.client", "GoogleCredentials"))}
```
Comment (garrettjonesgoogle) [inline vgen/src/main/resources/io/gapi/vgen/py/discovery_fragment.snip]: Doesn't addImport already return an empty string? Why do you need the silent() method?

File: vgen/src/test/java/io/gapi/vgen/py/PythonImportTest.java [hunk 6]
```diff
@@ -28,85 +28,85 @@
 
   @Test
   public void testImport_simple() {
-    PythonImport imp = PythonImport.create("foo", PythonImport.ImportType.STDLIB);
+    PythonImport imp = PythonImport.create(PythonImport.ImportType.STDLIB, "foo");
```
Comment (garrettjonesgoogle) [inline vgen/src/test/java/io/gapi/vgen/py/PythonImportTest.java]: Adding a static import of STDLIB would make this a little more concise.

File: vgen/src/main/java/io/gapi/vgen/py/PythonDiscoveryContext.java (hunk 2 above)
Comment (tcoffee-google) [inline vgen/src/main/java/io/gapi/vgen/py/PythonDiscoveryContext.java]: Done

File: vgen/src/main/java/io/gapi/vgen/py/PythonDiscoveryContext.java (hunk 3 above)
Comment (tcoffee-google) [inline vgen/src/main/java/io/gapi/vgen/py/PythonDiscoveryContext.java]: Done

File: vgen/src/test/java/io/gapi/vgen/testdata/discoveries/py/py_appengine.v1beta4.json.baseline (hunk 4 above)
Comment (tcoffee-google) [inline vgen/src/test/java/io/gapi/vgen/testdata/discoveries/py/py_appengine.v1beta4.json.baseline]: I was tempted, but this clarifies to the user that the terminal result of, e.g., list_next will be None (rather than, say, False), which may be helpful if they want to modify the handling.

File: vgen/src/main/resources/io/gapi/vgen/py/discovery_fragment.snip (hunk 5 above)
Comment (tcoffee-google) [inline vgen/src/main/resources/io/gapi/vgen/py/discovery_fragment.snip]: No, the naming is different in Python (from Java) to be consistent with what was done for Gapic. In Java, typeName returns names for referencing imports, and addImport silently wraps typeName. In Python, the "addImport" methods are analogous to typeName (to follow what was done before), and since there are multiple such methods, it seemed more convenient to have an independent silencing wrapper.

File: vgen/src/test/java/io/gapi/vgen/py/PythonImportTest.java (hunk 6 above)
Comment (tcoffee-google) [inline vgen/src/test/java/io/gapi/vgen/py/PythonImportTest.java]: Done

**Comment by tcoffee-google** [comment]
My commit amendments are showing up now, so I will leave as is for further review.

In the future, I will add new commits to the fork so that incremental changes are readily visible.

File: vgen/src/main/java/io/gapi/vgen/py/PythonLanguageProvider.java [hunk 7]
```diff
@@ -58,9 +58,10 @@
 
     Doc filenameDoc = snippets.generateFilename(element);
     String outputFilename = filenameDoc.prettyPrint();
+    Doc body = snippets.generateBody(element);
     List<String> importList = importHandler.calculateImports();
     // Generate result.
-    Doc result = snippets.generateClass(element, importList);
+    Doc result = snippets.generateModule(element, body, importList);
```
Comment (garrettjonesgoogle) [inline vgen/src/main/java/io/gapi/vgen/py/PythonLanguageProvider.java]: So the Python codegen (before your change) figures out all of the imports in an independent process from generating the class; it appears that you are converting it to work like the Java one now, where imports are accumulated as part of generating the body (in a stateful way). I don't have much opinion on whether one of these approaches should be used universally, but I wanted to call out this change in approach, because the authors of Python codegen may have picked their approach for a particular reason. @ethanbao @geigerj

Comment (geigerj) [inline vgen/src/main/java/io/gapi/vgen/py/PythonLanguageProvider.java]: Sorry, I haven't actually read through this whole change, but it looks like right here it's still the old flow of calling "calculateImports()" first, and then passing this list to the snippet entry point?

Comment (tcoffee-google) [inline vgen/src/main/java/io/gapi/vgen/py/PythonLanguageProvider.java]: The old process for Gapic library codegen is preserved: the corresponding generateBody snippet is degenerate. But we needed snippet-dependent imports to support proper sample gen for discovery (and, presumably, eventually for Gapic), so I changed the snippet interface to be like that for Java.

Comment (geigerj) [inline vgen/src/main/java/io/gapi/vgen/py/PythonLanguageProvider.java]: Ok, sounds good to me.

File: vgen/src/test/java/io/gapi/vgen/testdata/discoveries/java/java_appengine.v1beta4.json.baseline [hunk 8]
```diff
@@ -51,6 +51,8 @@ public class AppengineExample {
 
     Appengine.Apps.Get request = appengineService.apps().get(appsId);
     Application response = request.execute();
+
+    // TODO: Add code here to process the 'response' object
```
Comment (vchudnov-g) [inline vgen/src/test/java/io/gapi/vgen/testdata/discoveries/java/java_appengine.v1beta4.json.baseline]: Try to favor keeping the PR focused on a single language, including changes to other languages only when required by the changes you make here. Other enhancements should be in a separate PR.

This is good both for cognitive load of the reviewers and for using the github review system, which gets bogged down with many files.

File: vgen/src/test/java/io/gapi/vgen/NodeJSDiscoveryFragmentGeneratorTest.java [hunk 9]
```diff
@@ -1,3 +1,17 @@
+/* Copyright 2016 Google Inc
```
Comment (vchudnov-g) [inline vgen/src/test/java/io/gapi/vgen/NodeJSDiscoveryFragmentGeneratorTest.java]: Why is this Node file in this PR? Could we pull it into the Node PRs?

File: vgen/src/main/resources/io/gapi/vgen/py/discovery_fragment.snip [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+        2. Install python client library for Google APIs using 'pip install --upgrade google-api-python-client'
+        3. This sample uses Application Default Credentials for Auth. If not already done, install the gcloud CLI from 
+           https://cloud.google.com/sdk/ and run 'gcloud beta auth application-default login'
+        """
+    
+        @join import : imports
+            {@import}
+        @end
+        
+        @# Authentication is provided by gcloud tool when running locally
```
Comment (vchudnov-g) [inline vgen/src/main/resources/io/gapi/vgen/py/discovery_fragment.snip]: s/by/by the/
s/gcloud/"gcloud"/  (check whether we need to do this for other languages)

File: vgen/src/main/java/io/gapi/vgen/py/PythonImport.java [hunk 11]
```diff
@@ -38,27 +38,30 @@
   public abstract ImportType type();
 
   /*
-   * Create a Python import with the given module, attribute and local names.
+   * Create a Python import of the given category, specifying (attribute), (module, attribute), or
```
Comment (vchudnov-g) [inline vgen/src/main/java/io/gapi/vgen/py/PythonImport.java]: s/category/type/

File: vgen/src/main/java/io/gapi/vgen/py/PythonDiscoveryContext.java [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+    return pythonCommon.wrapIfKeywordOrBuiltIn(upperCamelToLowerUnderscore(typeName));
+  }
+
+  @Override
+  public String getMethodName(Method method) {
+    return getSimpleName(getRename(method.getName(), RENAMED_METHOD_MAP));
+  }
+
+  /**
+   * Returns a name for the type of a field of a type, substituting the given name when a native
```
Comment (vchudnov-g) [inline vgen/src/main/java/io/gapi/vgen/py/PythonDiscoveryContext.java]: nit: "for a type's field's type"

My own personal preference is here and in a bunch of the comments below to use a string of possessives to go from the outside in, rather than a string of "of"s to go from the inside out.

File: vgen/src/main/resources/io/gapi/vgen/py/discovery_fragment.snip (hunk 10 above)
Comment (tcoffee-google) [inline vgen/src/main/resources/io/gapi/vgen/py/discovery_fragment.snip]: Done

File: vgen/src/main/java/io/gapi/vgen/py/PythonImport.java (hunk 11 above)
Comment (tcoffee-google) [inline vgen/src/main/java/io/gapi/vgen/py/PythonImport.java]: Done

File: vgen/src/main/java/io/gapi/vgen/py/PythonDiscoveryContext.java (hunk 12 above)
Comment (tcoffee-google) [inline vgen/src/main/java/io/gapi/vgen/py/PythonDiscoveryContext.java]: Done

**Comment by tcoffee-google** [comment]
@vchudnov-g @pongad PTAL

**Comment by pongad** [comment]
This looks good to me

**Comment by garrettjonesgoogle** [comment]
For your commit message, remove the line about renaming the yaml files, that has already been done outside of this PR.

File: src/test/java/com/google/api/codegen/testdata/discoveries/py/py_appengine.v1beta4.json.baseline [hunk 13]
```diff
@@ -0,0 +1,472 @@
...
+   https://cloud.google.com/sdk/ and run 'gcloud beta auth application-default login'
+"""
+
+from __future__ import print_function
+import json
+
+from apiclient import discovery
+from oauth2client.client import GoogleCredentials
+
+# Authentication is provided by gcloud tool when running locally
```
Comment (vchudnov-g) [inline src/test/java/com/google/api/codegen/testdata/discoveries/py/py_appengine.v1beta4.json.baseline]: s/by gcloud tool/by the 'gcloud' tool/

File: src/test/java/com/google/api/codegen/testdata/discoveries/py/py_appengine.v1beta4.json.baseline [hunk 14]
```diff
@@ -0,0 +1,472 @@
...
+# and by built-in service accounts when running on GAE, GCE or GKE.
```
Comment (vchudnov-g) [inline src/test/java/com/google/api/codegen/testdata/discoveries/py/py_appengine.v1beta4.json.baseline]: nit: I personally prefer the Oxford comma, but up to you:
"GAE, GCE, or GKE."

File: src/test/java/com/google/api/codegen/testdata/discoveries/py/py_appengine.v1beta4.json.baseline [hunk 15]
```diff
@@ -0,0 +1,472 @@
...
+credentials = GoogleCredentials.get_application_default()
+
+# Construct the appengine service object (version v1beta4) for interacting
+# with the API. You can browse other available API services and versions at
+# https://developers.google.com/api-client-library/python/apis/
+service = discovery.build('appengine', 'v1beta4', credentials=credentials)
+
+
+# TODO: Change placeholders below to values for parameters to the 'get' method:
+# * Part of `name`. Name of the application to get. For example: "apps/myapp".
```
Comment (vchudnov-g) [inline src/test/java/com/google/api/codegen/testdata/discoveries/py/py_appengine.v1beta4.json.baseline]: You use back ticks here but regular ticks below in the TODO. I suggest regular quotes everywhere if that's possible.

**Comment by tcoffee-google** [comment]
@vchudnov-g PTAL

File: src/test/java/com/google/api/codegen/testdata/discoveries/py/py_appengine.v1beta4.json.baseline (hunk 13 above)
Comment (tcoffee-google) [inline src/test/java/com/google/api/codegen/testdata/discoveries/py/py_appengine.v1beta4.json.baseline]: AppEngine v1beta4 was vestigial --- removed

File: src/test/java/com/google/api/codegen/testdata/discoveries/py/py_appengine.v1beta4.json.baseline (hunk 14 above)
Comment (tcoffee-google) [inline src/test/java/com/google/api/codegen/testdata/discoveries/py/py_appengine.v1beta4.json.baseline]: Me too --- done

File: src/test/java/com/google/api/codegen/testdata/discoveries/py/py_appengine.v1beta4.json.baseline (hunk 15 above)
Comment (tcoffee-google) [inline src/test/java/com/google/api/codegen/testdata/discoveries/py/py_appengine.v1beta4.json.baseline]: Per discussion, deferring this issue pending further consideration.

**Comment by vchudnov-g** [comment]
LGTM
